src/main/go/org/jenkinsci/plugins/durabletask/heartbeat-launcher.go

```diff
@@ -0,0 +1,165 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"os/signal"
+	"sync"
+
+	"golang.org/x/sys/unix"
+)
+
+// Note: EXIT signal is needed as some shell variants require an EXIT trap after catching a signal
+const trapSig = "trap ':' INT TERM EXIT"
+
+func checkHeartbeatErr(err error) {
+	checkErr("heartbeat", err)
+}
+
+func checkLauncherErr(err error) {
+	checkErr("launcher", err)
+}
+
+func checkErr(process string, err error) {
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "(%v) check err: %v\n", process, err.Error())
+	}
+}
+
+// Catch a termination signal to allow for a graceful exit (i.e. no zombies)
+func signalCatcher(sigChan chan os.Signal) {
+	for sig := range sigChan {
+		fmt.Printf("(sig catcher) caught: %v\n", sig)
+	}
+}
+
+func launcher(wg *sync.WaitGroup, pidChan chan int,
+	scriptString string, logPath string, resultPath string, outputPath string) {
+
+	defer wg.Done()
+	recordExit := fmt.Sprintf("status=\"$?\"; echo \"$status\" > %v.tmp; mv %v.tmp %v; wait; exit \"$status\"",
+		resultPath, resultPath, resultPath)
+	scriptWithExit := trapSig + "; " + scriptString + "; " + recordExit
+	scriptCmd := exec.Command("/bin/sh", "-c", scriptWithExit)
+	logFile, err := os.Create(logPath)
+	checkLauncherErr(err)
+	defer logFile.Close()
+
+	if outputPath != "" {
+		// capturing output
+		outputFile, err := os.Create(outputPath)
+		checkLauncherErr(err)
+		defer outputFile.Close()
+		scriptCmd.Stdout = outputFile
+	} else {
+		scriptCmd.Stdout = logFile
+	}
+	if outputPath != "" {
+		// capturing output
+		scriptCmd.Stderr = logFile
+	} else {
+		// Note: pointing to os.Stdout will not capture all err logs and fail unit tests
```

Comment:
Do we have some test coverage verifying that something like

```groovy
sh 'echo problems here >&2'
```

indeed prints `problems here` to the build log?

Reply:
We have at least one in configuredinterpreter where you pass a bad interpreter and let it error out.